lib/utils/interval/multi.go

```diff
@@ -0,0 +1,397 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package interval
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/utils/retryutils"
+)
+
+// MultiInterval is equivalent to Interval except that it supports multiple intervals simultanesouly,
+// distinguishing them by key. The only real benefit to using this type instead of using multiple
+// Intervals is that it only allocates one timer and one background goroutine regardless of the number
+// of intervals. There are very few cases where this distinction matters. An example of a place where this
+// *does* matter is when you need multiple intervals *per* connected instance on an auth or proxy server.
+// In such a case, allocating only one timer and one goroutine per connected instance can be a significant
+// saving (see the lib/inventory for an example of this usecase).
+//
+// Note that MultiInterval behaves differently than time.Ticker or Interval in that it may yield the same
+// timestamp multiple times. It will only do this for *different* keys (i.e. K1 and K2 may both tick at T0)
+// but it is still a potential source of bugs/confusion when transitioning to using this type from one
+// of the single-interval alternatives.
+type MultiInterval[T comparable] struct {
+	subs      []subIntervalEntry[T]
+	push      chan subIntervalEntry[T]
+	ch        chan Tick[T]
+	reset     chan T
+	fire      chan T
+	closeOnce sync.Once
+	done      chan struct{}
+}
+
+// Tick represents a firing of the interval. The Key field denominates the sub-interval that
+// fired, and the Time field represents the time at which the firing occurred.
+type Tick[T any] struct {
+	Key  T
+	Time time.Time
+}
+
+// SubInterval configures an interval.  The only required parameters are the Duration
+// field which *must* be a positive duration, and the Key field which can be any comparable
+// value.
+type SubInterval[T any] struct {
```

Comment:
I'm not sure what the go best practices are, but in my previous generics experience, it's always been best practice to use the broadest constraint possible for each type.  Since there is no hard requirement for `Tick` or `Subinterval` to impose `comparable`, we don't artificially apply the constraint.  Preserves maximum flexibility.

Say, for example, you wanted to write a function that prints ticks.  If `Tick[T any]` then all you need to do is this:

```
func PrintTick[T fmt.Stringer](tick Tick[T]) { ... }
```

If `Tick[T comparable]`, then you need to write a custom interface just to make this function work:

```
type comparableStringer {
    comparable
    fmt.Stringer
}

func PrintTick[T comparableStringer](tick Tick[T]) { ... }
```

Basically, generic constraints are "contagious".  Once you add a constraint to one type, it starts to creep into all the types that interact with that type.  The solution to this problem is to write the constraint as late as possible, so that only the types that *really* need the constraint to function actually apply it.